### PR TITLE
Non interactive state initialization

### LINKED ISSUE #2389: Bug with terragrunt run-all init --terragrunt-non-interactive does not work
Hi,

We have a use case where we want to pass y by default to the prompt to create the backend bucket - we have tried using the command 

terragrunt run-all init --terragrunt-non-interactive

But the above command gives a 409 error during initializing saying bucket already exists after its created and init isnt successful

We also tried to non-interactive commands like 

echo "Y" | terragrunt run-all init

echo "y" | terragrunt run-all init

echo "YES" | terragrunt run-all init

echo "yes" | terragrunt run-all init

But none of the work, they end up giving errors, but although the bucket does get created in all the above cases, the initialization fails with errors in all cases

For apply, we have used the same command echo "Y" | terragrunt run-all apply which is working perfectly fine for us for the prompt do you want to run terragrunt in all modules? (y/n)

But the same command does not work for creating the bucket

What we noticed is we get 2 prompts for creating buckets, and we are not sure if that is causing issues when we pass y through non-interactive mode

The request for 2 prompts looks like a bug and maybe this is why non-interactive mode isn't working either where non-interactive provides y for first bucket prompt and creates it, and again provides y for the second prompt too and throws error saying bucket already exists

I am not sure if this is a bug

Terragrunt version AMD - 46
terraform version - 1.3.6

vasudevaner-macbookpro:modules vasudevaner$ terragrunt run-all init
INFO[0000] The stack at /Users/vasudevaner/Documents/Data/EDW_E2E/edw-e2e/Foundation/modules will be processed in the following order for command init:
Group 1
- Module /Users/vasudevaner/Documents/Data/EDW_E2E/edw-e2e/Foundation/modules
- Module /Users/vasudevaner/Documents/Data/EDW_E2E/edw-e2e/Foundation/modules/gcs

Group 2
- Module /Users/vasudevaner/Documents/Data/EDW_E2E/edw-e2e/Foundation/modules/pubsub

Group 3
- Module /Users/vasudevaner/Documents/Data/EDW_E2E/edw-e2e/Foundation/modules/gcc

Group 4
- Module /Users/vasudevaner/Documents/Data/EDW_E2E/edw-e2e/Foundation/modules/cloudrun

Group 5
- Module /Users/vasudevaner/Documents/Data/EDW_E2E/edw-e2e/Foundation/modules/bq
 
Remote state GCS bucket edw-test-tfhcltest-state does not exist or you don't have permissions to access it. Would you like Terragrunt to create it? (y/n) Remote state GCS bucket edw-test-tfhcltest-state does not exist or you don't have permissions to access it. Would you like Terragrunt to create it? (y/n)

### PR DESCRIPTION
**Pull request by denis256**
## Description

Included changes:
* added locks per bucket name to allow only one initialization process to run for bucket
* added AWS and GCP tests to track that initialization is not failing

Fixes #2389.

## TODOs

Read the [Gruntwork contribution guidelines](https://gruntwork.notion.site/Gruntwork-Coding-Methodology-02fdcd6e4b004e818553684760bf691e).

- [x] Update the docs.
- [x] Run the relevant tests successfully, including pre-commit checks.
- [x] Ensure any 3rd party code adheres with our [license policy](https://www.notion.so/gruntwork/Gruntwork-licenses-and-open-source-usage-policy-f7dece1f780341c7b69c1763f22b1378) or delete this line if its not applicable.
- [x] Include release notes. If this PR is backward incompatible, include a migration guide.

## Release Notes (draft)

Added / Removed / Updated [X].

Updated bucket initialization to avoid concurrency errors.

### CODE REVIEW

**Review by brikis98 — COMMENTED** [review]
Thanks for this PR!

A few high level thoughts:

1. This does locking within one TG process. So if someone runs `run-all xxx`, this will ensure that all the goroutines in there don't try to create/modify the same backend at the same time. That seems useful 👍 
2. This doesn't help with locking across processes. So if two separate devs run TG on the same backend that doesn't yet exist, those two may hit concurrency issues. I think when we originally wrote at least the S3 code, it was possible to detect that someone else (in the same process or not) was creating or modifying the same bucket concurrently based on the error that came back from AWS (after all, they handle concurrency themselves already). Any reason we can't still do that? That would be even more powerful...

File: remote/remote_state.go [hunk 1]
```diff
@@ -215,3 +224,26 @@ type BucketCreationNotAllowed string
...
+	return &stateAccess{
+		m: make(map[string]*sync.Mutex),
+	}
+}
+
+func (om *stateAccess) Do(key string, f func() error) error {
+	om.mu.Lock()
+	defer om.mu.Unlock()
+
+	if om.m == nil {
```
[the remark this answers is not in the document]
Comment (denis256) [reply remote/remote_state.go]: Updated variable names

File: remote/remote_state.go [hunk 2]
```diff
@@ -215,3 +224,26 @@ type BucketCreationNotAllowed string
...
+func newStateAccess() *stateAccess {
+	return &stateAccess{
+		m: make(map[string]*sync.Mutex),
+	}
+}
+
+func (om *stateAccess) Do(key string, f func() error) error {
+	om.mu.Lock()
+	defer om.mu.Unlock()
+
```
[the remark this answers is not in the document]
Comment (denis256) [reply remote/remote_state.go]: I updated approach in 2 functions:
fetchMutex() - return mutex for specific bucket, it has inside mutex for map access
StateBucketUpdate() - execute logic for state update, using obtained mutex from fetchMutex()

in this way will be ensured that co-routines will be blocked per bucket name.

**Review by brikis98 — APPROVED** [review]
I think this will fix issues with `run-all` within one process, so let's get that merged.

That said, I still think if we handle error responses from AWS APIs better, we could also handle issues with concurrency across multiple processes... But that can be done later.

**Pull request merged**